pkg/pillar/cmd/zedrouter/zedrouter.go

```diff
@@ -1618,6 +1720,84 @@ func doAppNetworkModifyUnderlayNetwork(
 	publishNetworkInstanceStatus(ctx, netstatus)
 
 	maybeRemoveStaleIpsets(staleIpsets)
+	return nil
+}
+
+// returns a interfaces list, with network change
+func doAppNetworkModifyUNetAppNums(
+	ctx *zedrouterContext,
+	config types.AppNetworkConfig,
+	oldConfig types.AppNetworkConfig,
+	status *types.AppNetworkStatus,
+	ipsets []string) ([]bool, error) {
+	var intfList []bool
+	for i := range config.UnderlayNetworkList {
+		log.Tracef("handleModify ulNum %d\n", i)
+		ulConfig := &config.UnderlayNetworkList[i]
+		oldulConfig := &oldConfig.UnderlayNetworkList[i]
+		ulStatus := &status.UnderlayNetworkList[i]
+		change, err := doAppNetworkModifyUNetAppNum(ctx,
+			status, ulConfig, oldulConfig, ulStatus, ipsets)
+		if err != nil {
+			return intfList, err
+		}
+		intfList = append(intfList, change)
+	}
+	publishAppNetworkStatus(ctx, status)
+	return intfList, nil
+}
+
+// on network transition,
+// if activated, deactivate the underlay interface
+// release the appNum and acquire appNum on the new network instance
+func doAppNetworkModifyUNetAppNum(
+	ctx *zedrouterContext,
+	status *types.AppNetworkStatus,
+	ulConfig *types.UnderlayNetworkConfig,
+	oldulConfig *types.UnderlayNetworkConfig,
+	ulStatus *types.UnderlayNetworkStatus,
+	ipsets []string) (bool, error) {
+	appID := status.UUIDandVersion.UUID
+	networkID := ulConfig.Network
+	oldNetworkID := ulStatus.Network
+	// no change, return
+	if uuid.Equal(networkID, oldNetworkID) {
+		return false, nil
+	}
+	netstatus := lookupNetworkInstanceStatus(ctx, ulConfig.Network.String())
+	if netstatus == nil {
+		errStr := fmt.Sprintf("network instance status get fail :%s",
+			networkID.String())
+		log.Errorf("doAppNetworkModifyUNetAppNum(%s, %s): fail: %s",
+			networkID.String(), appID.String(), errStr)
+		return false, errors.New(errStr)
+	}
+	oldnetstatus := lookupNetworkInstanceStatus(ctx, oldNetworkID.String())
+	if oldnetstatus == nil {
+		errStr := fmt.Sprintf("network instance status get fail :%s",
+			oldNetworkID.String())
+		log.Errorf("doAppNetworkModifyUNetAppNum(%s, %s): fail: %s",
+			networkID.String(), appID.String(), errStr)
+		return false, errors.New(errStr)
+	}
+	// release ACLs, IP, Vifs and other resources on the old network
+	if status.Activated {
+		appNetworkDoInactivateUnderlayNetwork(ctx, status, ulStatus, ipsets)
+	}
+	// release the app number on old network
+	if _, err := appNumOnUNetGet(ctx, oldNetworkID, appID); err == nil {
+		appNumOnUNetFree(ctx, oldNetworkID, appID)
+	}
+	// allocate an app number on new network
+	isStatic := (ulConfig.AppIPAddr != nil)
+	if _, err := appNumOnUNetAllocate(ctx, networkID, appID,
+		isStatic, false); err != nil {
+		errStr := fmt.Sprintf("App Num get fail :%s", err)
+		log.Errorf("appNumsOnUNetAllocate(%s, %s): fail: %s",
+			networkID.String(), appID.String(), errStr)
+		return false, errors.New(errStr)
+	}
+	return true, nil
```

Comment:
There is something very assymetric about this function.

If uuid.Equal(networkID, oldNetworkID), then it leaves things in the same state, in particular if it was activated it is still activated.
But if the networkid has changed, then it returns with it inactivated. Can't the function be symmetric and do the activate in the case of a change? That would make things much easier to understand.